src/porepy/numerics/ad/discretizations.py

```diff
@@ -479,166 +501,292 @@ def _differentiable_mpfa(
+        bound_pressure_face_jac = sps.csr_matrix(
+            (base_bound_pressure_face.shape[0], self.dof_manager.num_dofs())
+        )
+
+        projections = pp.ad.SubdomainProjections(self.grids)
+
+        # The differentiation of transmissibilities with respect to permeability is
+        # implemented as a loop over all grids. It could be possible to gather all grid
+        # information in arrays as a preprocessing step, but that seems not to be worth
+        # the effort. However, to avoid evaluating the permeability function multiple
+        # times, we precompute it here
+        global_permeability = perm_function(perm_argument).evaluate(self.dof_manager)
+
+        g: pp.Grid
+        for g in self.grids:
+            params = self.gb.node_props(g)[pp.PARAMETERS][self.keyword]
+            transmissibility_jac, inverse_sum_squared = self._transmissibility(
+                g, global_permeability
+            )
+
+            # On Dirichlet faces, the tpfa discretization simply recovers boundary
+            # condition. Thus, no t-differentiation enters.
+            # On Neumann faces, however, the tpfa contribution is
+            # v_face[bnd.is_neu] = -1 / t_full[bnd.is_neu]
+            # Differentiate and multiply with transmissibility jacobian
+            v_face = np.zeros(g.num_faces)
+            is_neu = params["bc"].is_neu
+
+            v_face[is_neu] = inverse_sum_squared[is_neu]
+            d_face2boundp_d_t = sps.diags(v_face, shape=(g.num_faces, g.num_faces))
+            jac_bound_pressure_face = d_face2boundp_d_t * transmissibility_jac
+
+            # Prolong this Jacobian to the full set of faces and add.
+            face_prolongation = projections.face_prolongation([g]).evaluate(
+                self.dof_manager
+            )
+            bound_pressure_face_jac += face_prolongation * jac_bound_pressure_face
+
+        bound_pressure = pp.ad.Ad_array(
+            base_bound_pressure_face, bound_pressure_face_jac
+        )
+        return bound_pressure
+
+    def _geometry_information(
+        self, g
+    ) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+        """Utility function to retrieve geometry information.
+
+        Args:
+            g: pp.Grid for which the information is extracted.
+
+        Returns:
+            Arrays representing face indices, cell indices, normal vector signs and vectors
+            between face centers and cell centers. The former three have size
+                    n = num_face_cell_pairs = g.num_face * 2 (two cells per face),
+            while fc_cc.shape = (3, n).
+        """
+        fi, ci, sgn = sps.find(g.cell_faces)
+
+        # Distance from face center to cell center
+        fc_cc = g.face_centers[::, fi] - g.cell_centers[::, ci]
+        return fi, ci, sgn, fc_cc
+
+    def _transmissibility(
+        self, g, global_permeability: pp.ad.Ad_array
+    ) -> sps.csr_matrix:
```

Comment:
```suggestion
    ) -> Union[sps.csr_matrix, np.ndarray]:
```
Unsure about the second argument, but there should at least be a second argument (mypy will insist)